operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakIngress.java

```diff
@@ -0,0 +1,140 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.operator.v2alpha1;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.networking.v1.IngressBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.api.model.networking.v1.Ingress;
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.OperatorManagedResource;
+import org.keycloak.operator.StatusUpdater;
+import org.keycloak.operator.v2alpha1.crds.Keycloak;
+import org.keycloak.operator.v2alpha1.crds.KeycloakStatusBuilder;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+public class KeycloakIngress extends OperatorManagedResource implements StatusUpdater<KeycloakStatusBuilder> {
+
+    private Ingress existingIngress;
+    private final Keycloak keycloak;
+
+    public KeycloakIngress(KubernetesClient client, Keycloak keycloakCR) {
+        super(client, keycloakCR);
+        this.keycloak = keycloakCR;
+        this.existingIngress = fetchExistingIngress();
+    }
+
+    private <T> boolean checkField(Ingress ingress1, Ingress ingress2, Function<Ingress, T> selector) {
+        try {
+            return selector.apply(ingress1).equals(selector.apply(ingress2));
+        } catch (NullPointerException npe) {
+            return false;
+        }
+    }
+
+    @Override
+    protected Optional<HasMetadata> getReconciledResource() {
+        var defaultIngress =newIngress();
+        if (keycloak.getSpec().isDefaultIngressDisabled() && existingIngress != null) {
+            client.network().v1().ingresses().delete(existingIngress);
+            return Optional.empty();
+        } else if (existingIngress == null) {
+            return Optional.of(defaultIngress);
+        } else if (
+                !checkField(existingIngress, defaultIngress, i -> i.getMetadata().getAnnotations().get("nginx.ingress.kubernetes.io/backend-protocol")) ||
+                !checkField(existingIngress, defaultIngress, i -> i.getSpec().getDefaultBackend().getService().getPort().getNumber()) ||
+                !checkField(existingIngress, defaultIngress, i -> i.getSpec().getRules().get(0).getHttp().getPaths().get(0).getBackend().getService().getPort().getNumber()) ||
+                !checkField(existingIngress, defaultIngress, i -> i.getSpec().getRules().get(0).getHost()) ||
+                !checkField(existingIngress, defaultIngress, i -> i.getSpec().getRules().get(0).getHttp().getPaths().get(0).getBackend().getService().getName()) ||
+                !checkField(existingIngress, defaultIngress, i -> i.getSpec().getDefaultBackend().getService().getName())
+        ) {
+            existingIngress.getMetadata().getAnnotations().putAll(existingIngress.getMetadata().getAnnotations());
+            existingIngress.setSpec(defaultIngress.getSpec());
```

Comment:
I believe that we're effectively rewriting the whole specs, right? Not just the few selected one that are managed by the operator (protocol, port, host, service name)?

Reply:
that's correct.